From 9a5f14cad91173255b5a2ec2f88735b850db55e5 Mon Sep 17 00:00:00 2001
From: Jayachandran C <jchandra@broadcom.com>
Date: Mon, 16 Jun 2014 16:05:40 +0530
Subject: sata: Reset SATA interface after u-boot

U-boot leaves SATA interface in a partially initialized way when
SATA disks are used from u-boot. Reset the interface from Linux with
the SATA configuration.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/xlp/ahci-init-xlp2.c b/arch/mips/netlogic/xlp/ahci-init-xlp2.c
index e1af277..f4894cc 100644
--- a/arch/mips/netlogic/xlp/ahci-init-xlp2.c
+++ b/arch/mips/netlogic/xlp/ahci-init-xlp2.c
@@ -259,6 +259,14 @@ static void nlm_sata_firmware_init(int node)
 	pr_info("Initializing XLP9XX On-chip AHCI...\n");
 	regbase = nlm_get_sata_regbase(node);
 
+	/* Work around for u-boot leaving SATA in in-determinate mode */
+	reg_val = nlm_read_sata_reg(regbase, SATA_CTL);
+	if (reg_val != 0x800800acU) {
+		pr_info("Hard reset AHCI...\n");
+		nlm_write_sata_reg(regbase, SATA_CTL, 0x800800acU);
+		udelay(1000);
+	}
+
 	/* Reset port0 */
 	sata_clear_glue_reg(regbase, SATA_CTL, P0_IRST_POR);
 	sata_clear_glue_reg(regbase, SATA_CTL, P0_IRST_HARD_TXRX);
-- 
1.7.1

